main.go

```diff
@@ -0,0 +1,226 @@
+		chanEntry.bundleName = fields[2]
+		chanEntry.depth, _ = strconv.Atoi(fields[3])
+		chanEntry.bundleVersion = fields[4]
+		chanEntry.bundleSkipRange = fields[5]
+		chanEntry.replacesBundleName = fields[6]
+
+		if pkgToGraph != "" && chanEntry.packageName != pkgToGraph {
+			continue
+		}
+		// Get or create package
+		p, ok := pkgs[chanEntry.packageName]
+		if !ok {
+			p = &pkg{
+				name:    chanEntry.packageName,
+				bundles: make(map[string]*bundle),
+			}
+		}
+		pkgs[chanEntry.packageName] = p
+
+		// Get or create bundle
+		bundl, ok := p.bundles[chanEntry.bundleName]
+		if !ok {
+			bundl = &bundle{
+				name:              chanEntry.bundleName,
+				packageName:       chanEntry.packageName,
+				minDepth:          chanEntry.depth,
+				isBundlePresent:   chanEntry.bundleVersion != "",
+				channels:          sets.NewString(),
+				replaces:          sets.NewString(),
+				skipRangeReplaces: sets.NewString(),
+			}
+			if chanEntry.bundleSkipRange != "" {
+				bundl.skipRange = chanEntry.bundleSkipRange
+			}
+			if chanEntry.bundleVersion != "" {
+				bundl.version = chanEntry.bundleVersion
+			} else {
+				// catch empty version field because empty '()' are a Mermaid syntax error
+				bundl.version = "x.y.z"
+			}
+		}
+		p.bundles[chanEntry.bundleName] = bundl
+
+		bundl.channels.Insert(chanEntry.channelName)
+		if chanEntry.replacesBundleName != "" {
+			bundl.replaces.Insert(chanEntry.replacesBundleName)
+		}
+		if chanEntry.depth < bundl.minDepth {
+			bundl.minDepth = chanEntry.depth
+		}
+	}
+	for _, p := range pkgs {
+		for _, pb := range p.bundles {
+			if pb.skipRange == "" {
+				continue
+			}
+			pSkipRange, err := semver.ParseRange(pb.skipRange)
+			if err != nil {
+				fmt.Errorf("invalid range %q for bundle %q: %v", pb.skipRange, pb.name, err)
+				continue
+			}
+			for _, cb := range p.bundles {
+				if !cb.isBundlePresent {
+					continue
+				}
+				cVersion, err := semver.Parse(cb.version)
+				if err != nil {
+					fmt.Errorf("invalid version %q for bundle %q: %v", cb.version, cb.name, err)
+					continue
+				}
+				if pSkipRange(cVersion) {
+					pb.skipRangeReplaces.Insert(cb.name)
+				}
+			}
+		}
+	}
+
+	return pkgs, nil
+}
+
+func outputMermaidScript(pkgs map[string]*pkg) {
+	graphHeader()
+	for _, pkg := range pkgs {
+		allBundleChannels := sets.NewString()                     // we want subgraph organized by channel
+		fmt.Fprintf(os.Stdout, "\n"+indent1+"subgraph "+pkg.name) // per package graph
+		for _, bundle := range pkg.bundles {
+			allBundleChannels = bundle.channels.Union(allBundleChannels)
+		}
+		for _, channel := range allBundleChannels.List() {
+			fmt.Fprintf(os.Stdout, "\n"+indent2+"subgraph "+channel+" channel") // per channel graph
+			replaceSet := sets.NewString()
+			for _, bundle := range pkg.bundles {
+				if bundle.channels.Has(channel) {
+					// if no replaces edges, just write the node
+					if bundle.replaces.Len() == 0 && bundle.skipRangeReplaces.Len() == 0 {
```

Comment:
We have skip and skipRange. What is this one?

Reply:
Maybe the naming is bad? `skipRangeReplaces` is a set composed here:
https://github.com/operator-framework/index-mermaid-graph/pull/1/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R157

For example when processing the `jaeger-product` package:
`pbSkipRange` is the semver range parsed from `>=1.13.0 <2.0.0-tp.1` and in a positive instance, `jaeger-operator.v1.17.8` is included in the set. This part of the post-processing logic is untouched from here:

https://github.com/joelanford/olm-index-graph/commit/76e13a46f801d483989204b47ce16198788da3c8#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R192

It could still not be what we want now, but that's where it derives from.

We use that set to determine where to draw graph edges connecting nodes (versions) that are allowed to replace other versions due to the `skipRange` in the CSV, the ones we draw in Mermaid with ` o--o `

Reply:
However, we have 2 ways to set up a skip:
1 - olm.skips (list of bundles )
2 - olm.skipRange (range )

So, note that they have diff legend as well (why that is important? the olm.skips remove the bundle from the catalog the other not and diff issues/scenarios can be faced). Then see: 

<img width="832" alt="Screen Shot 2021-03-19 at 15 44 23" src="https://user-images.githubusercontent.com/7708031/111806703-2b877100-88ca-11eb-94bd-0c19f0bfa591.png">
<img width="798" alt="Screen Shot 2021-03-19 at 15 44 30" src="https://user-images.githubusercontent.com/7708031/111806710-2cb89e00-88ca-11eb-988e-194afa257057.png">


Are we showing the legends accordingly in these cases?

Reply:
Ah, I will need to look closer at properly accounting for skips. Thanks!

Reply:
Thank you.

Reply:
To follow up on `skips`, now that handling for them is included in the code. For the following result set for `advance-cluster-management operator`:

| pkgName | chanName | bndlName | bndl-Depth | bndl-Version | bndlSkipRange | skips | replaceBndlName | head | default_chan |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.7|0|2.0.7|""|""|advanced-cluster-management.v2.0.6|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.6|1|2.0.6|""|""|advanced-cluster-management.v2.0.5|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.5|2|2.0.5|""|""|advanced-cluster-management.v2.0.4|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.4|3|2.0.4|""|""|advanced-cluster-management.v2.0.3|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.3|4|2.0.3|""|""|advanced-cluster-management.v2.0.2|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.2|5|2.0.2|""|""|advanced-cluster-management.v2.0.1|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.1|6|2.0.1|""|""|advanced-cluster-management.v2.0.0|advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.0|advanced-cluster-management.v2.0.0|7|2.0.0|""|""||advanced-cluster-management.v2.0.7|release-2.1
advanced-cluster-management|release-2.1|advanced-cluster-management.v2.1.2|0|2.1.2|>=2.0.0 <2.1.0|advanced-cluster-management.v2.1.1|advanced-cluster-management.v2.1.0|advanced-cluster-management.v2.1.2|release-2.1
advanced-cluster-management|release-2.1|advanced-cluster-management.v2.1.1|1|2.1.1|>=2.0.0 <2.1.0|""||advanced-cluster-management.v2.1.2|release-2.1
advanced-cluster-management|release-2.1|advanced-cluster-management.v2.1.2|1|2.1.2|>=2.0.0 <2.1.0|advanced-cluster-management.v2.1.1|advanced-cluster-management.v2.1.1|advanced-cluster-management.v2.1.2|release-2.1
advanced-cluster-management|release-2.1|advanced-cluster-management.v2.1.0|2|2.1.0|>=2.0.0 <2.1.0|""||advanced-cluster-management.v2.1.2|release-2.1

Now yields these graphs:

<img width="1648" alt="image" src="https://user-images.githubusercontent.com/4449199/114582685-d2072d80-9c4e-11eb-9843-60be947de43f.png">

Reply:
I think we can close this comment as solved.